grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy

```diff
@@ -9,29 +9,144 @@
+import org.hibernate.criterion.DetachedCriteria
+import org.hibernate.criterion.Projections
+import org.hibernate.criterion.Restrictions
+import org.hibernate.criterion.Subqueries
+import org.hibernate.sql.JoinType
 import org.pih.warehouse.core.Organization
 import org.pih.warehouse.core.PreferenceType
 import org.pih.warehouse.core.ProductPrice
 import org.pih.warehouse.core.RatingTypeCode
 import org.pih.warehouse.core.UnitOfMeasure
-import org.pih.warehouse.importer.ImportDataCommand
 import org.pih.warehouse.product.Product
 import org.pih.warehouse.product.ProductPackage
 import org.pih.warehouse.product.ProductSupplier
+import org.pih.warehouse.product.ProductSupplierListDto
+import org.pih.warehouse.product.ProductSupplierListParams
 import org.pih.warehouse.product.ProductSupplierPreference
-import util.ConfigHelper
 
 import java.text.SimpleDateFormat
 
 @Transactional
 class ProductSupplierService {
 
+    public static final PREFERENCE_TYPE_NONE = "NONE"
+
+    public static final PREFERENCE_TYPE_MULTIPLE = "MULTIPLE"
+
     def identifierService
     def dataSource
 
+
+    List<ProductSupplierListDto> getProductSuppliers(ProductSupplierListParams params) {
+        // Store added aliases to avoid duplicate alias exceptions for product and supplier
+        // This could happen when params.searchTerm and e.g. sort by productCode/productName is applied
+        Set<String> usedAliases = new HashSet<>()
+
+        List<ProductSupplier> productSuppliers =
+            ProductSupplier.createCriteria().list(max: params.max, offset: params.offset) {
+                if (params.searchTerm) {
+                    createAlias("product", "p", JoinType.LEFT_OUTER_JOIN)
+                    createAlias("supplier", "s", JoinType.LEFT_OUTER_JOIN)
+                    usedAliases.addAll(["product", "supplier"])
+                    or {
+                        ilike("p.productCode", "%" + params.searchTerm + "%")
+                        ilike("code", "%" + params.searchTerm + "%")
+                        ilike("s.code", "%" + params.searchTerm + "%")
+                    }
+                }
+                if (params.product) {
+                    eq("product.id", params.product)
+                }
+                if (params.active != null) {
+                    eq("active", params.active)
+                }
+                if (params.supplier) {
+                    eq("supplier.id", params.supplier)
+                }
+                if (params.preferenceType) {
+                    add(getPreferenceTypeCriteria(params.preferenceType))
+                }
+                if (params.createdFrom) {
+                    ge("dateCreated", params.createdFrom)
+                }
+                if (params.createdTo) {
+                    lte("dateCreated", params.createdTo)
+                }
+                if (params.sort) {
+                    String orderDirection = params.order ?: "asc"
+                    switch (params.sort) {
```

Comment:
Let's move this to a method if possible, as it is going to need a lot of unit tests
```
Map sortOrder = getSortOrder(params.sort, params.order, usedAliases)
order(sortOrder.property, sortOrder.direction)
```

NOTE: I'm not confident with the naming of the properties in the sortOrder map, so if you think there's a more intuitive naming convention please let me know.

Reply:
I will try if it is possible to do it in an easy way. I mean - I'd have to search for a package that would have the order static method available. It seems not to be available in the `Restrictions`.
I see there is a class `Order` from `org.hibernate.Criterion` so I will eventually try this one.